akka-stream/src/main/scala/akka/stream/impl/ActorMaterializerImpl.scala

```diff
@@ -244,17 +241,31 @@ private[akka] class SubFusingActorMaterializerImpl(
 
   def takeSnapshotsOfChildren(): Future[immutable.Seq[StreamSnapshot]] = {
     // Arbitrary timeout but should always be quick, the failure scenario is that
-    // the child/stream stopped, and we do retry below
-    implicit val timeout: Timeout = 1.second
-    def takeSnapshot() = {
-      val futureSnapshots =
-        context.children.toList.map(child => (child ? ActorGraphInterpreter.Snapshot).mapTo[StreamSnapshot])
-      Future.sequence(futureSnapshots)
-    }
-
+    // the child/stream stopped, and we do retry below.
     // If the timeout hits it is likely because one of the streams stopped between looking at the list
     // of children and asking it for a snapshot. We retry the entire snapshot in that case
-    retry(() => takeSnapshot(), 3)
+    val children = context.children.toList
+    def takeSnapshot(): Future[List[StreamSnapshot]] = {
+      implicit val ec: ExecutionContext = ExecutionContexts.parasitic
+      val futureSnapshots = children.map(child => {
+        val futureSnapshotOfChild = RetrySupport.retry(() => {
+          implicit val timeout: Timeout = 1.second
+          (child ? ActorGraphInterpreter.Snapshot).mapTo[StreamSnapshot]
+        }, 3)
+        futureSnapshotOfChild.transform {
+          case Success(snapshot)  => Success((child, Success(snapshot)))
+          case Failure(exception) => Success((child, Failure(exception)))
+        }
+      })
+      Future
+        .sequence(futureSnapshots)
+        .map(_.foldLeft(List.newBuilder[StreamSnapshot]) {
+          case (builder, (_, Success(snapshot))) => builder += snapshot
+          case (builder, (child, Failure(exception))) =>
+            if (child.isTerminated) builder else throw exception
```

Comment:
Only fixed the thread-safety problem for now.

Reply:
More detailed implementation will need to introduce a new sub type of `StreamSnapshot`.

Reply:
Are you going to follow this up soon? Otherwise, it would be good if could add more information about which child failed here directly, maybe just with an extra log line.

Reply:
I will do more during the 5.1 holiday, I think because of it's internal API, the better way should be include all unexpected child to the dump. how do you think about?

Reply:
When I ran into it, it failed because one of the actors below the supervisor was a TlsActor which didn't support snapshot. However, since it was named "flow-2-1` or something that wasn't apparent. Would be nice that whatever we do, these kind of things become more obvious.